### PR TITLE
Add support for AGP 3.1.x

### PR DESCRIPTION
**Pull request by raviagarwal7**
- Bump AGP to `3.1.0-beta3`
- Bump gradle to `4.5.1`
- Bump support to `27.0.2`
- Use butterknife version `9.0.0-SNAPSHOT`
- use build tools `27.0.3`
- Bump compile and target sdk to `27`
- Fix travis

Does not work with AGP 3.0.x as some apis used by okbuck were internal in those releases

### CODE REVIEW

**Comment by runningcode** [comment]
Also, looks like the builds have been failing for a while, but we'll probably need to bump the sdk versions in the travis build files.

**Comment by kageiit** [comment]
@raviagarwal7 Can you take a stab at fixing travis while at it :)

File: transform-cli/src/main/java/com/uber/okbuck/transform/TransformInvocationBuilder.java [hunk 1]
```diff
@@ -146,5 +147,15 @@ public File getTemporaryDir() {
         public String getPath() {
             throw new UnsupportedOperationException();
         }
+
+        @Override
```
Comment (kageiit) [inline transform-cli/src/main/java/com/uber/okbuck/transform/TransformInvocationBuilder.java]: cc @malbano

File: build.gradle [hunk 2]
```diff
@@ -4,6 +4,8 @@ buildscript {
         jcenter()
         google()
         maven { url "https://plugins.gradle.org/m2/" }
+        // Remove once butterknife is updated to a stable version
```
Comment (cwoodwar6) [inline build.gradle]: Make a TODO and link a ticket to fix?

File: build.gradle [hunk 3]
```diff
@@ -20,6 +22,9 @@ allprojects { project ->
     repositories {
         jcenter()
         google()
+        maven { url "https://plugins.gradle.org/m2/" }
+        // Remove once butterknife is updated to a stable version
```
Comment (cwoodwar6) [inline build.gradle]: Same here, TODO and a ticket to remove

File: buildSrc/build.gradle [hunk 4]
```diff
@@ -62,7 +64,7 @@ task sourcesJar(type: Jar) {
     classifier = 'sources'
 }
 
-def publishVersion = '0.32.1'
+def publishVersion = '0.32.1-uber'
```
Comment (cwoodwar6) [inline buildSrc/build.gradle]: Internal version name? :)

File: build.gradle (hunk 2 above)
Comment (raviagarwal7) [reply build.gradle]: done

File: build.gradle (hunk 3 above)
Comment (raviagarwal7) [reply build.gradle]: done

File: buildSrc/build.gradle (hunk 4 above)
Comment (raviagarwal7) [reply buildSrc/build.gradle]: renamed to `0.33.0-alpha01`

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy [hunk 5]
```diff
@@ -52,8 +50,8 @@ class AndroidInstrumentationTarget extends AndroidAppTarget {
     Scope getInstrumentation() {
         return Scope.from(
                 project,
-                expand(compileConfigs, ANDROID_TEST_PREFIX)
-                        + ["androidTest${getMainTargetName(name).capitalize()}Compile"],
+                ImmutableSet.of(baseVariant.runtimeConfiguration.name,
+                        ["androidTest${getMainTargetName(name).capitalize()}RuntimeClasspath"]),
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidInstrumentationTarget.groovy]: lets remove the second line

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 6]
```diff
@@ -134,12 +133,40 @@ abstract class AndroidTarget extends JavaLibTarget {
     Scope getTest() {
         return Scope.from(
                 project,
-                expand(compileConfigs, TEST_PREFIX, true),
+                unitTestVariant ?
+                        ImmutableSet.of(
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: can we make a new constuctor in scope to directly accept a configuration

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy [hunk 7]
```diff
@@ -409,32 +436,14 @@ abstract class AndroidTarget extends JavaLibTarget {
...
-        List<String> expansions = ["", buildType, flavor, name]
-        Set<String> expandedConfigs = configNames.collect { String configName ->
-            expansions.collect { String expansion ->
-                String expanded = [prefix, expansion.capitalize(), configName.capitalize()].join('')
-                expanded.substring(0, 1).toLowerCase() + expanded.substring(1)
-            }
-        }.flatten() as Set<String>
-        if (prefix && includeParent) {
-            expandedConfigs += expand(configNames)
+    protected Set<String> expand(Set<String> configNames, String prefix = "") {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/android/AndroidTarget.groovy]: unused looks like

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaTarget.groovy [hunk 8]
```diff
@@ -1,84 +1,74 @@
...
 
 abstract class JavaTarget extends JvmTarget {
 
-    protected static final String TEST_PREFIX = "test"
+    protected static final String UNIT_TEST_PREFIX = "test"
 
-    private static final Set<String> JAVA_COMPILE_CONFIGS = ["compile"]
-    private static final Set<String> JAVA_APT_CONFIGS = ["apt"]
-    public static final Set<String> JAVA_PROVIDED_CONFIGS = ["compileOnly", "provided"]
+    public static final Set<String> JAVA_DEPS_CONFIGS = ["runtimeClasspath"]
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaTarget.groovy]: Lets use the constants from https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaTarget.groovy [hunk 9]
```diff
@@ -1,84 +1,74 @@
...
+    private static final Set<String> JAVA_PROVIDED_DEPS_CONFIGS = ["compileClasspath"]
+    private static final Set<String> JAVA_APT_CONFIGS = ["apt", "annotationProcessorClasspath"]
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaTarget.groovy]: Lets try to remove `apt` so we can use `annotationProcessor` even for `java-library`

File: buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java [hunk 10]
```diff
@@ -88,4 +100,45 @@ static String findVersionInClasspath(Project project, String group, String modul
                 .map(ModuleComponentIdentifier::getVersion)
                 .orElse(null);
     }
+
+
+    // Taken from AGP 3.1.0 ArtifactDependencyGraph
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java]: Taken - > Copied

Comment (kageiit) [reply buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java]: Also include the license header in the docs

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 11]
```diff
@@ -96,6 +98,30 @@ class Scope {
         }).flatten() as Set<String>).findAll { !it.empty }
     }
 
+    private static Set<ResolvedArtifactResult> getArtifacts(
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: can you add some javadoc here

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 12]
```diff
@@ -108,24 +134,19 @@ class Scope {
...
         artifacts.each { ResolvedArtifactResult artifact ->
             if (!DependencyUtils.isConsumable(artifact.file)) {
                 return
             }
             ComponentIdentifier identifier = artifact.id.componentIdentifier
             if (identifier instanceof ProjectComponentIdentifier) {
-                targetDeps.add(ProjectUtil.getTargetForOutput(project.project(identifier.projectPath), artifact.file))
+
+                String variant = artifact.variant.attributes.getAttribute(VariantAttr.ATTRIBUTE)
+                targetDeps.add(ProjectUtil.getTargetForOutput(
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: rename to `getTargetForVariant`

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 13]
```diff
@@ -108,24 +134,19 @@ class Scope {
...
-        Set<ResolvedArtifactResult> artifacts = configurations.collect {
-            it.incoming.artifacts.artifacts
-        }.flatten() as Set<ResolvedArtifactResult>
+        Set<ResolvedArtifactResult> artifacts = getArtifacts(configurations)
 
         artifacts.each { ResolvedArtifactResult artifact ->
             if (!DependencyUtils.isConsumable(artifact.file)) {
                 return
             }
             ComponentIdentifier identifier = artifact.id.componentIdentifier
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: Lets collect these identifiers here in some collection as we are iterating already (if we need to download sources later)

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 14]
```diff
@@ -154,5 +175,13 @@ class Scope {
                 external.add(greatest.get(it).first())
             }
         }
+
+        if (project.rootProject.okbuck.intellij.sources) {
+            Set<ComponentIdentifier> ids = artifacts.collect { artifact ->
+                artifact.getId().getComponentIdentifier()
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: collect in the loop above if we need to download sources

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/TargetCache.java [hunk 15]
```diff
@@ -89,15 +87,15 @@
     }
 
     @Nullable
-    public Target getTargetForOutput(Project targetProject, File artifact) {
+    public Target getTargetForOutput(Project targetProject, String variant) {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/TargetCache.java]: getTargetForVariant

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 16]
```diff
@@ -54,17 +56,27 @@ class Scope {
         jvmArgs = jvmArguments
         this.depCache = depCache
 
-        extractConfigurations(configurations)
+        extractConfiguration(configuration)
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: lets guard against null here since we already have the deps sets initialized to be empty

File: buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyUtils.groovy [hunk 17]
```diff
@@ -13,19 +13,19 @@ final class DependencyUtils {
...
-        configurations.each { String configName ->
-            try {
-                useful.add(project.configurations.getByName(configName))
-            } catch (UnknownConfigurationException ignored) {}
-        }
+    static Configuration useful(Project project, String configuration) {
+        try {
+            Configuration config = project.configurations.getByName(configuration)
+            return useful(config)
+        } catch (UnknownConfigurationException ignored) {}
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyUtils.groovy]: return null in the catch block and change `ignored` to `e`

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 18]
```diff
@@ -54,17 +56,27 @@ class Scope {
...
+    }
+
+    static Scope from(Project project,
+                      Configuration configuration,
+                      Set<File> sourceDirs = ImmutableSet.of(),
+                      @Nullable File resDir = null,
+                      List<String> jvmArguments = ImmutableList.of(),
+                      DependencyCache depCache = ProjectUtil.getDependencyCache(project)) {
+        Configuration useful = DependencyUtils.useful(configuration)
+        String key = useful ? useful.name : "n/a"
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: lets change "n/a" to "None"

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaLibTarget.groovy [hunk 19]
```diff
@@ -15,20 +15,28 @@ class JavaLibTarget extends JavaTarget {
...
                 project.file("src/main/resources"),
                 project.compileJava.options.compilerArgs as List)
     }
 
     @Override
     Scope getTest() {
         return Scope.from(project,
-                expand(compileConfigs, TEST_PREFIX, true),
-                project.sourceSets.test.java.srcDirs as Set,
+                expand(depsConfig, UNIT_TEST_PREFIX),
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaLibTarget.groovy]: we can remove this expanding behavior since we can ask the java plugin for the singular configuration name similar to what we do for android

File: buildSrc/src/main/java/com/uber/okbuck/core/task/OkBuckCleanTask.java [hunk 20]
```diff
@@ -79,4 +79,4 @@ public String getDescription() {
     public String getGroup() {
         return OkBuckGradlePlugin.GROUP;
     }
-}
+}
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/core/task/OkBuckCleanTask.java]: can we undo this change?

File: buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java [hunk 21]
```diff
@@ -88,4 +100,59 @@ static String findVersionInClasspath(Project project, String group, String modul
...
+            // where sources are not available
+            Set<ComponentIdentifier> remainingToResolve = Sets.newHashSet();
+            for (ComponentArtifactsResult componentResult : resolvedComponents) {
+                Set<ArtifactResult> sourcesArtifacts =
+                        componentResult.getArtifacts(SourcesArtifact.class);
+                if (sourcesArtifacts.isEmpty()) {
+                    remainingToResolve.add(componentResult.getId());
+                }
+            }
+            if (!remainingToResolve.isEmpty()) {
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java]: we dont care about javadoc artifacts, so lets remove this logic. We only care about sources

Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/task/OkBuckCleanTask.java (hunk 20 above)
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/task/OkBuckCleanTask.java]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaLibTarget.groovy (hunk 19 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/java/JavaLibTarget.groovy]: removed for ones which have constants in JavaPlugin. still need to keep it for apt.

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy (hunk 18 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: changed it to `--none--` just in case someone keeps a `None` configuration

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy (hunk 16 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyUtils.groovy (hunk 17 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/dependency/DependencyUtils.groovy]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java [hunk 22]
```diff
@@ -88,4 +100,41 @@ static String findVersionInClasspath(Project project, String group, String modul
...
+            query.forComponents(artifacts);
+
+            @SuppressWarnings("unchecked")
+            Class<? extends Artifact>[] artifactTypesArray =
+                    (Class<? extends Artifact>[]) new Class<?>[] {SourcesArtifact.class};
+            query.withArtifacts(JvmLibrary.class, artifactTypesArray);
+            ArtifactResolutionResult queryResult = query.execute();
+            queryResult.getResolvedComponents();
+        } catch (Throwable t) {
+            System.out.println("Unable to download sources/javadoc for project " +
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/core/util/ProjectUtil.java]: Unable to download `sources`

not `sources/javadoc`

**Comment by artem-zinnatullin** [comment]
But what about mapping `implementation` dependencies to proper `VISIBILITY` in `BUCK` files?
https://buckbuild.com/concept/visibility.html

If I'm not missing something, currently OkBuck hardcodes `VISIBILITY` as `PUBLIC`.

Without proper `VISIBILITY` mapping between `build.gradle` and `BUCK` configs we can see different compilation results in Gradle vs Buck. 

ie dependency that is declared as `implementation` will prevent Gradle from compiling another module that tries to use that dependency as transitive, while in Buck it'll work which will be a real issue on big projects

**Comment by kageiit** [comment]
> ie dependency that is declared as implementation will prevent Gradle from compiling another module that tries to use that dependency as transitive, while in Buck it'll work which will be a real issue on big projects

Hey @artem-zinnatullin this is not the case as okbuck handles this by ensuring that the deps declared in the buck files are not visible to consumers that depend on them. We use `deps` and not `exported_deps` which is a very important distinction. `VISIBILITY` is not in the picture here. We expose exactly what gradle exposes on the compileclasspath and also what ends up in the apk.

**Comment by artem-zinnatullin** [comment]
@kageiit I'll have to double check that, not sure I immediately understand how does that protect us from implicit transitive deps across modules

Other than that strong +1 from me on merge 👍, I'm applying it on our master branch today, thank you a lot for pushing PR and publishing alpha! ❤️

**Comment by artem-zinnatullin** [comment]
AGP `3.2.0-alpha01` was released, releases from this branch seem to work fine with it :)

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 23]
```diff
@@ -96,45 +104,92 @@ class Scope {
...
-        Set<ResolvedArtifactResult> artifacts = configurations.collect {
-            it.incoming.artifacts.artifacts
-        }.flatten() as Set<ResolvedArtifactResult>
+        return artifactResults
+    }
 
+    private void extractConfiguration(Configuration configuration) {
         Set<ComponentIdentifier> artifactIds = new HashSet<>()
+
+        Spec<ComponentIdentifier> filter = new Spec<ComponentIdentifier>() {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: extract to private static field and add javadoc

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 24]
```diff
@@ -96,45 +104,92 @@ class Scope {
...
                 return
             }
+
+            ProjectComponentIdentifier identifier = artifact.id.componentIdentifier
+            String variant = artifact.variant.attributes.getAttribute(VariantAttr.ATTRIBUTE)
+            targetDeps.add(ProjectUtil.getTargetForVariant(
+                    project.project(identifier.projectPath), variant))
+        }
+
+        filter = new Spec<ComponentIdentifier>() {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: same

File: buildSrc/src/main/java/com/uber/okbuck/core/util/LintUtil.java [hunk 25]
```diff
@@ -56,15 +57,23 @@ private static String getLintwConfigName(Project project, File config) {
...
+
     @SuppressWarnings("ResultOfMethodCallIgnored")
     public static synchronized String getLintwConfigRule(Project project, File config) {
-        File configFile = project.getRootProject().file(LINT_DEPS_CACHE + "/" + getLintwConfigName(project, config));
-        try {
-            FileUtils.copyFile(config, configFile);
-        } catch (IOException e) {
-            throw new RuntimeException(e);
+        String configName =  getLintwConfigName(project, config);
+
```
Comment (kageiit) [inline buildSrc/src/main/java/com/uber/okbuck/core/util/LintUtil.java]: add a comment

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 26]
```diff
@@ -37,6 +37,26 @@ class Scope {
     protected final Project project
     final Set<ExternalDependency> external = new HashSet<>()
 
+    /**
+     * Used to filter out only project dependencies when resolving a configuration.
+     */
+    private static Spec<ComponentIdentifier> projectFilter = new Spec<ComponentIdentifier>() {
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: make final and capitalize name

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy [hunk 27]
```diff
@@ -37,6 +37,26 @@ class Scope {
...
+        boolean isSatisfiedBy(ComponentIdentifier componentIdentifier) {
+            return componentIdentifier instanceof ProjectComponentIdentifier
+        }
+    }
+
+    /**
+     * Used to filter out external & local jar/aar dependencies when resolving a configuration.
+     */
+    private static Spec<ComponentIdentifier> externalDepFilter = new Spec<ComponentIdentifier>() {
+        @Override
```
Comment (kageiit) [inline buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: make final and capitalize name

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy (hunk 26 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: done

File: buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy (hunk 27 above)
Comment (raviagarwal7) [reply buildSrc/src/main/groovy/com/uber/okbuck/core/model/base/Scope.groovy]: done

File: libraries/javalibrary/build.gradle [hunk 28]
```diff
@@ -1,14 +1,16 @@
-apply plugin: 'java'
-apply plugin: "net.ltgt.apt"
+apply plugin: 'java-library'
```
[the remark this answers is not in the document]
Comment (kageiit) [reply libraries/javalibrary/build.gradle]: java-library applies the java plugin as well

https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaLibraryPlugin.java#L37
